### PR TITLE
Implement #127 and #96

### PR DESCRIPTION
**Pull request by davidlatwe**
### What's Changed

* Application package version can be changed in main view (#96)
* Version list has been limited to profile request range (#127)

### Internal Refactor

* ~`PackagesModel` and `ApplicationModel` has been merged into one `ResolvedPackagesModel`.~
* ~Two proxy model were added to adopt the above merge : `ApplicationProxyModel` and `PackagesProxyModel`.~
* Controller is no longer required for resetting package model.

### Discussion

After package version list has been limited to only show in the range that profile requested, the `latest` flag is not reflecting *real* latest anymore.

But do we need to know whether a package is latest or not ? I do imaging that sometime we indeed want to know that info, but should not be often. So I would propose to implement that query action in a menu, and only scan all versions when one needed.

Secondly, do we still need the "Set to earliest" and "Set to latest" action ?
<details><summary> action snapshot </summary>

image

</details>

I am more lean to respect the request what profile have made, and not allowing change denpency version freely.

What do you think ?

### CODE REVIEW

**Comment by davidlatwe** [comment]
Ah, one more thing, the "Beta".
Since the version string already contains "beta", do we still need package model to parse that info ?
<details><summary>snapshot</summary>

image

</details>

**Comment by davidlatwe** [comment]
Was looking at the development journal [#LatestVersion](https://gist.github.com/mottosso/90e28ab935f20aea7d7bce95c7969b41#latest-version), and it was enough for me to keep the _Beta_ column. :P

The reason for me to brought that up was only try to make model data lesser.

About the _Latest_, I think maybe we could make a background thread to collect that info after model rested.

**Comment by mottosso** [comment]
> But do we need to know whether a package is latest or not ?

I think so? Especially as you develop I would think? And especially as someone else updates a package you aren't working with directly.

But how do you mean "real" latest? What else is it showing?

> Secondly, do we still need the "Set to earliest" and "Set to latest" action ?

We need to make it an option at least, something that can be configured. Potentially something visible to developers, but not to artists.

**Comment by davidlatwe** [comment]
> But how do you mean "real" latest? What else is it showing?

The implementation of limiting version list is to find package versions with fixed range, which leads to only showing the latest one of that range.

Shouldn't be hard to solve, and with the input you gave, I think I have a rough UI/UX plan for _overriding_ versions now. 🤔

First would be to implement a background thread to fetch a full version list for each package in context, after profile resolved with all applications. Once collected, all packages can override version freely if user ticked the checkbox e.g. "Unlock version range".

**Comment by davidlatwe** [comment]
Maybe, one extra function would be, auto bump or down other dependencies version if current version changing package leads to a failed resolve.

File: allzpark/control.py [hunk 1]
```diff
@@ -1193,9 +1208,23 @@ def _try_resolve_context(req, pkg_name, mode):
             if hidden and not show_hidden:
                 continue
 
-            visible_apps += [package]
+            versions = rez.find(package.name,
+                                range_=app_ranges[package.name],
```
Comment (davidlatwe) [inline allzpark/control.py:1212]: Version limiting with profile request happens here.

**Comment by RafaelVillar** [comment]
Hello @davidlatwe ! The work you have been doing is fantastic!

In regards to the "real latest" question. I think the latest of "range" is an okay assumption to make, no? I understand it may not be the literal latest of <package name> but if it was an artist is allowed then that should be ok. I think this feature originally sprang from a discussion of convenience to quickly return something back to default or if you are developer testing packages, you would be incrementing.

**Comment by mottosso** [comment]
(For context, Allzpark was initially developed for Rafael's studio. He'll understand the pre-existing requirements well)

**Comment by davidlatwe** [comment]
Thanks @RafaelVillar and @mottosso  !

> I understand it may not be the literal latest of but if it was an artist is allowed then that should be ok.

Agree :)

So I end up with adding "Show All Versions" in preference.

When that enabled, dependency package will be able to select all versions from repository, except applications that has fixed range set by profile. This should reduce the chance of selecting version that can't be resolved with profile.

I also make packages that has only one version available not triggering version editing delegate widget, and set bold font for package that has many. This should make user much easier to know which is changeable.

And since "Show All Versions" is mostly for developer, I also implemented a way to lock preference with default value from `allzparkconfig.py`.

Lastly, I found that graph button will be disabled even package is not missing but only conflicted, which is a bug from #124, so I fixed it together in this PR.

lock

File: allzpark/control.py [hunk 2]
```diff
@@ -1181,25 +1200,50 @@ def _try_resolve_context(req, pkg_name, mode):
...
-            visible_apps += [package]
+            app_versions = [str(v.version) for v in app_ranges[app_pkg.name]]
+            resolved_packages = {
+                pkg: [str(p.version)
+                      for p in (self.find(pkg.name) if all_vers else [pkg])]
+                for pkg in contexts[request].resolved_packages or []
+                if pkg.name not in [app_pkg.name, profile_variant.name]
+            }
+            # profile version should not be changed in Packages view
+            resolved_packages[profile_variant] = [profile_variant]
```
Comment (davidlatwe) [inline allzpark/control.py:1229]: All package versions are getting picked here if "Show All Versions" is enabled. No background thread stuff, thought to keep it simple.

Comment (mottosso) [reply allzpark/control.py:1229]: So long as it keeps the UI responsive. No locking up at any time.

Comment (davidlatwe) [reply allzpark/control.py:1229]: Yeah, if there are really massive versions to collect, will only increase profile load time, once being loaded, no extra loading between application selection change.

**Comment by davidlatwe** [comment]
I found packages' resolve order become incorrect in "Packages" dock widget after the change that collecting all packages and versions in one place with merged model.

So I went back and after reviewing my changes a few times, I decide to revert the `PackagesModel` and `ApplicationMode` merge, mainly for two reasons.

One is to fix the package resolve ordering of course, and the other is, I started to worry that not only reset time will become longer, but also it will be hard to implement feature like "not loading resolved packages when 'Packages' dock page is not visible/active".

I think this is a good revert.

Oh and, this PR also fixes #87.

**Comment by davidlatwe** [comment]
After adding version editing test cases, I found changing preference "Show All Versions" in test will get segfault easily.

So I had a look and found it was because the preference widget (qargparse widgets) were defined as `Preferences` dock widget's class attribute (not defined in `__init__`), which makes them not able to be garbage collected when tearing down the test.

Somehow, those preference widgets that survived from previous tests still able to touch models that are created in current running test, and how many test has ran, how many model rest may called, when changing preference that calls `ctrl.reset`, e.g. "Show All Versions", hence segfault.

However, although after commit feed9c3 resolved the problem in new test case (tested with my Windows and WSL Ubuntu), test running on Azure DevOp still have a little chance getting segfault. The journey hasn't end.

**Comment by davidlatwe** [comment]
Ok, I think this is stable now, here what we have now :

* App package version can be changed in main view (#96)
* App's version list has been limited to profile request range (#127)
* Right-clicking on any column in Packages view can bring up version editor (#87)
* By default, only app that has multi-versions can change/patch version in main Apps view or in Packages page
* If preference "Show All Versions" enabled, all package versions will be listed, except profile and app package
* App selection won't lost (jump to first app in view) after patching
* App version won't be editable if versions has been flattened ([allzpark.com#multiple-application-versions](https://allzpark.com/gui/#multiple-application-versions))
* Preference setting can be blocked from `allzparkconfig.protected_preferences` to prevent touching danger stuff
* No `TableModel` needs `Controller` object to get packages now (tech-debt paid)
* A few tests added for this PR

And here's the changes in action :

(Since dependency package in `allzparkdemo` only `core_pipeline` has multiple versions, so only that package can be edit when "Show All Version" enabled)

latest

Ready to merge now :)

**Comment by mottosso** [comment]
Great work, David. :)

**Comment by davidlatwe** [comment]
Merging this later today :)

**Pull request merged**